### PR TITLE
refactor: ktor 3 compatibility

### LINKED ISSUE #136: Ktor 3 compatibility
Although there are [no official release notes](https://github.com/ktorio/ktor/releases) yet, Ktor 3.0.0 final was published and tools like Renovate and Dependabot start updating projects to it.

During that upgrade, our project tests now [fail with](https://github.com/eclipse-apoapsis/ort-server/actions/runs/11259267196/job/31307842354?pr=1188#step:5:246)

```
     java.lang.IncompatibleClassChangeError: Found interface io.ktor.server.routing.Route, but class was expected
        at io.github.smiley4.ktorswaggerui.dsl.routing.DocumentedRouteSelectorKt.documentation(DocumentedRouteSelector.kt:41)
        at io.github.smiley4.ktorswaggerui.dsl.routing.DocumentedRouteSelectorKt.get(DocumentedRouteSelector.kt:99)
        at org.eclipse.apoapsis.ortserver.core.api.HealthKt.healthChecks(Health.kt:30)
        at org.eclipse.apoapsis.ortserver.core.plugins.RoutingKt.configureRouting$lambda$2$lambda$1(Routing.kt:39)
        at io.ktor.server.routing.RoutingBuilderKt.route(RoutingBuilder.kt:21)
```

So looks like this project is affected by some breaking changes in Ktor 3.

### PR DESCRIPTION
**Pull request by sahitono**
KTOR 3 Compatibility will close #136 , there are many breaking change to the Route of ktor.
There are issue in my environment that i can not run CustomizedSchemaGenerator.kt as it says i need kotlin 2.0 to run kotlinx 1.7.3. I am not sure is that happen only at my environment because i use kotlin 2.0

### CODE REVIEW

**Comment by sahitono** [comment]
image

it looks like need to change the kotlinx serialization to 1.7.3 because ktor 3 is using that. but that means it need to be kotlin 2.0, what do you think ? should it be postponed or create a new branch  @SMILEY4

**Comment by SMILEY4** [comment]
(without having tested or looked anything up) i dont see any issue with upgrading kotlinx-serialization. Im just thinking if we should also maybe test this with https://github.com/SMILEY4/schema-kenerator first. If there could be any issues, its will be there.

**Comment by SMILEY4** [comment]
I did some quick testing schema-kenerator and this project and i don't think upgrading kotlin (2.0.21?) and serialization (1.7.3) should be an issue :+1: 
I would move the schema-kenerator project to those versions aswell.

**Comment by sahitono** [comment]
got it, i will update both and the schema kenerator

**Comment by SMILEY4** [comment]
Thank you. I already started it for schema-kenerator https://github.com/SMILEY4/schema-kenerator/compare/upgrade-to-kotlin-2. I dont think there is much more to do.
Is there a reason you chose kotlin version "2.0.20" over "2.0.21" ?

File: gradle.properties [hunk 1]
```diff
@@ -16,12 +16,12 @@ projectDeveloperName=smiley4
...
 
 # dependency versions
-versionKtor=2.3.12
+versionKtor=3.0.0
 versionSwaggerUI=5.17.11
 versionSwaggerParser=2.1.22
 versionSchemaKenerator=1.4.3
 versionKotlinLogging=3.0.5
 versionKotest=5.8.0
 versionKotlinTest=1.8.21
```
Comment (ShiinaKin) [inline gradle.properties]: just a suggest, keep the version of `kotlin test` aligned with kotlin

Comment (sahitono) [reply gradle.properties]: got it, bump it to 2.0.20

**Comment by sahitono** [comment]
> Thank you. I already started it for schema-kenerator [SMILEY4/schema-kenerator@`upgrade-to-kotlin-2` (compare)](https://github.com/SMILEY4/schema-kenerator/compare/upgrade-to-kotlin-2). I dont think there is much more to do. Is there a reason you chose kotlin version "2.0.20" over "2.0.21" ?

no, my mistake i thought 2.0.20 is the latest. i will bump it

File: gradle.properties (hunk 1 above)
Comment (mcxinyu) [reply gradle.properties]: The maintenance of the features of ktor2.+ may continue for some time, and it is recommended to wait until the next major version (for this project) before abandoning ktor 2.+ .

**Comment by mcxinyu** [comment]
I think we should not develop the compatibility of ktor3.0 in the develop branch, or at least create a new module for the compatibility of ktor3.0. Otherwise, this modification will affect other users who use ktor2.+.

**Comment by mcxinyu** [comment]
The maintenance of the features of ktor2.+ may continue for some time, and it is recommended to wait until the next major version (for this project) before abandoning ktor 2.+ .

File: ktor-swagger-ui-examples/build.gradle.kts [hunk 2]
```diff
@@ -5,7 +5,7 @@ version = projectVersion
 
 plugins {
     kotlin("jvm")
-    kotlin("plugin.serialization") version "1.9.25"
+    kotlin("plugin.serialization") version "2.0.20"
```
Comment (ShiinaKin) [inline ktor-swagger-ui-examples/build.gradle.kts]: serialization's version also could be update to `2.0.21` 😃

Comment (sahitono) [reply ktor-swagger-ui-examples/build.gradle.kts]: thank you for finding it, already bumped it

**Review by ShiinaKin — APPROVED** [review]

**Comment by SMILEY4** [comment]
> I think we should not develop the compatibility of ktor3.0 in the develop branch, or at least create a new module for the compatibility of ktor3.0. Otherwise, this modification will affect other users who use ktor2.+.

Agreed, my current plan would be to create one more minor version with some small features/improvements and then split off some branch for the maintenance of this project with ktor 2.x.

File: ktor-swagger-ui/src/main/kotlin/io/github/smiley4/ktorswaggerui/dsl/routing/resources/DocumentedResourceRoutes.kt [hunk 3]
```diff
@@ -4,19 +4,16 @@ import io.github.smiley4.ktorswaggerui.dsl.routes.OpenApiRoute
...
-import io.ktor.util.pipeline.PipelineContext
 
 //============================//
 //             GET            //
 //============================//
 
 inline fun <reified T : Any> Route.get(
     noinline builder: OpenApiRoute.() -> Unit = { },
-    noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+    noinline body: suspend io.ktor.server.routing.RoutingContext.() -> Unit
```
Comment (sschuberth) [inline ktor-swagger-ui/src/main/kotlin/io/github/smiley4/ktorswaggerui/dsl/routing/resources/DocumentedResourceRoutes.kt:16]: Out of curiosity, why the FQN here, is there a conflict with another class of the same name?

Comment (sahitono) [reply ktor-swagger-ui/src/main/kotlin/io/github/smiley4/ktorswaggerui/dsl/routing/resources/DocumentedResourceRoutes.kt:16]: do you mean the RoutingContext ?

Comment (sschuberth) [reply ktor-swagger-ui/src/main/kotlin/io/github/smiley4/ktorswaggerui/dsl/routing/resources/DocumentedResourceRoutes.kt:16]: Yes.

**Pull request merged**